### PR TITLE
build: change ng_package "bundle index" files to non-conflicting paths

### PR DESCRIPTION
**Pull request by alexeagle**
Currently packages like @angular/core ship with the generated metadata
in a path like 'core.js' which overwrites one of the inputs.

Angular material puts the generated file in a path like 'index.js'

Either way these files generated by ng_module rules have the potential
to collide with inputs given by the user, which results in an error.

Instead, we have the freedom to choose a different non-conflicting name.

### CODE REVIEW

File: packages/bazel/src/ng_package/packager.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ function main(args: string[]): number {
     if (!rel) {
       rel = '.';
     }
-    const basename = nameParts[nameParts.length - 1];
+    const basename = nameParts[nameParts.length - 1] + '_public_index';
```
Comment (jelbourn) [inline packages/bazel/src/ng_package/packager.ts]: Add comment here that explains why we append `_public_index`? I would personally also add the same comment to the bzl file.

Separately we should make sure this gets some kind of note in the "spec".

**Review by jelbourn — COMMENTED** [review]
The generated `d.ts` at the root of the package (or deeper for nested entry-points) needs to also point to this new file name; it's currently still referencing e.g. `http/http` (line 147)

File: packages/bazel/src/ng_package/packager.ts (hunk 1 above)
Comment (alexeagle) [reply packages/bazel/src/ng_package/packager.ts]: done, and requested comment/edit access on the spec https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/edit

**Comment by alexeagle** [comment]
Fixed the nested entry point, see the golden file.

File: packages/bazel/src/ng_module.bzl [hunk 2]
```diff
@@ -253,46 +253,53 @@ def _ts_expected_outs(ctx, label):
...
   tsconfig = dict(tsc_wrapped_tsconfig(ctx, ctx.files.srcs, ctx.files.srcs), **{
     "angularCompilerOptions": {
-      "flatModuleOutFile": basename,
+      "flatModuleOutFile": flat_module_out_file,
     },
   })
-  if ctx.attr.module_name:
-    tsconfig["angularCompilerOptions"]["flatModuleId"] = ctx.attr.module_name
+  if not ctx.attr.module_name:
+    fail("Only ng_module with a module_name attribute should be exposed as flat module")
```
Comment (jelbourn) [inline packages/bazel/src/ng_module.bzl]: Expand this to explain why?

File: packages/bazel/src/ng_module.bzl [hunk 3]
```diff
@@ -253,46 +253,53 @@ def _ts_expected_outs(ctx, label):
   return _expected_outs(ctx)
 
 def _write_bundle_index(ctx):
```
Comment (jelbourn) [inline packages/bazel/src/ng_module.bzl]: Add function description for this that explains why these re-export files are written?

File: packages/bazel/src/ng_package/ng_package.bzl [hunk 4]
```diff
@@ -264,7 +264,12 @@ def _ng_package_impl(ctx):
     if f.path.endswith(".js"):
       esm2015.append(struct(js = f, map = None))
 
-  for entry_point in [""] + ctx.attr.secondary_entry_points:
+  entry_points = []
```
Comment (jelbourn) [inline packages/bazel/src/ng_package/ng_package.bzl]: Add comment that explains the criteria for which `deps` are considered entry points?

File: packages/bazel/src/ng_package/ng_package.bzl [hunk 5]
```diff
@@ -306,43 +311,62 @@ def _ng_package_impl(ctx):
         config_name = entry_point.replace("/", "_"))
     bundles.append(struct(js = min_output, map = uglify_sourcemap))
 
-  metadata_files = depset(transitive = [getattr(dep, "angular").flat_module_metadata
-                                        for dep in ctx.attr.deps
-                                        if hasattr(dep, "angular")])
+  inputs = (
```
Comment (jelbourn) [inline packages/bazel/src/ng_package/ng_package.bzl]: More descriptive name than `inputs`?

File: packages/bazel/src/ng_package/ng_package.bzl [hunk 6]
```diff
@@ -306,43 +311,62 @@ def _ng_package_impl(ctx):
...
 
   # The order of arguments matters here, as they are read in order in packager.ts.
   args.add(npm_package_directory.path)
   args.add(ctx.label.package)
-  args.add(primary_entry_point_name(ctx.attr.name, ctx.attr.entry_point))
-  args.add(ctx.attr.secondary_entry_points, join_with=",")
   args.add([ctx.bin_dir.path, ctx.label.package], join_with="/")
-  args.add(ctx.file.readme_md.path if ctx.file.readme_md else "")
+
+  flat_module_metadata = depset(transitive = [
```
Comment (jelbourn) [inline packages/bazel/src/ng_package/ng_package.bzl]: I don't quite follow what this does; add comment?

**Comment by alexeagle** [comment]
@jelbourn thanks for the feedback - I'm already working on a follow-up PR so I'll improve the comments as you suggest in there - since it takes a while to get this green again

**Review by IgorMinar — APPROVED** [review]
I love the tests! thanks

**Pull request closed without merging**